File: 39 - Combination Sum/solution.js
/**
 * @param {number[]} candidates
 * @param {number} target
 * @return {number[][]}
 *
 * Time Complexity - O((2^k) . k)
 * Space Complexity - O(k .x)
 * where x is number of combinations
 * 
 */
var combinationSum = function(candidates, target) {
    let result = [];

    const helper = (nums, target, ind, temp) => {
        if(ind == nums.length){
            if(target == 0){
                result.push([...temp]);
            }
            return;
        }

        if(nums[ind] <= target){
            temp.push(nums[ind]);
            helper(nums, target-nums[ind], ind, temp);
            temp.pop();
        }

        helper(nums, target, ind+1, temp);
    }

    helper(candidates, target, 0, []);
    return result;
};

File: 39 - Combination Sum/solution.java
import java.util.ArrayList;
import java.util.List;

class Solution {

    public void helper(int[] nums, int target, int ind, List<Integer> temp, List<List<Integer>> result){
        if(ind == nums.length){
            if(target == 0){
                result.add(new ArrayList<>(temp));
            }
            return;
        }

        if(nums[ind] <= target){
            temp.add(nums[ind]);
            helper(nums, target-nums[ind], ind, temp, result);
            temp.remove(temp.size()-1);
        }

        helper(nums, target, ind+1, temp, result);
    }

    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        List<List<Integer>> result = new ArrayList<>();

        helper(candidates, target, 0, new ArrayList<>(), result);

        return result;
    }
}